### ISSUE TITLE
F1-F12 keys support

### ISSUE THREAD

**Issue by jcorbalan** [intersystems-ru/webterminal#81]
I have opened this issue so you can to answer  for use of the F1 to F12 keys. Thank you.

**Comment by ZitRos** [comment]
@jcorbalan, thanks for opening the issue.

I have edited the issue so it now contains technical details to be implemented. Please confirm that this is correct.

**Comment by jcorbalan** [comment]
yes, it's correct.

**Comment by ZitRos** [comment]
@jcorbalan, one more fact I need to realize, does your terminal application work with Caché TERM? Maybe you use SSH clients like Putty to work with it. I cannot find out how to read the function keys in Caché TERM with one of these:

```
read *char
read string
```

If you have any suggestions, please comment. Thanks!

**Comment by jcorbalan** [comment]
No, I work with TUN PLUS 2016 http://www.esker.com/host-access-software/tun.asp

**Comment by jcorbalan** [comment]
The same problem is with arrow keys, don't work in webterminal, it's necesary
 	Reset	Set
up	 EscA	EscOA
down EscB       EscOB
right	 EscC	EscOC
left	EscD 	EscOD

**Comment by ZitRos** [comment]
I understood, thank you very much for clarifying this. I will get back in 1-2 days to this issue implementing required escape sequences, and then asking you to test the results (as I don't know any open solutions on M/COS languages to test this on).

**Comment by jcorbalan** [comment]
OK, No problem, i test it.
Another drawback are the graphic character, you can see here:
 http://www.theasciicode.com.ar/extended-ascii-code/graphic-character-low-density-dotted-ascii-code-176.html
they are the characteres from 176 to 223 . Now they are not shown in webterminal. 
The sequence of escape necessary to activate or deactivate them is Esc(0	Esc)0 that Set G0 special chars. & line.
 Please, yo can see :  http://ascii-table.com/ansi-escape-sequences-vt-100.php

**Comment by ZitRos** [comment]
Thanks! The idea is clear for now, let's deal first with function keys.

**Comment by jcorbalan** [comment]
You can use this command in TERM:, to get the value of the PF key pressed, which will be stored in the variable PF.
READ E9#10 
SET PF=$S($T=0:1,$ZB?.N:$ZB\256-31,$E($ZB,4)'=""&($A($ZB,4)'=126):$S($A($ZB,4)-49<5:$A($ZB,4)-39,1:$A($ZB,4)-49),$E($ZB,3)'="":$A($ZB,3)-79,1:-31)

**Comment by ZitRos** [comment]
OK, @jcorbalan, you are welcome to test function and arrow keys support. Please install the attached WebTerminal version (it is not released yet). Here are the [key mappings](https://github.com/intersystems-ru/webterminal/blob/keyMappings/src/client/js/input/keyMappings.js). They work as follows: both `read *c` and `read c` commands take mapped escape sequence to the variable `c`.

However, I haven't been able to 

> get the value of the PF key pressed, which will be stored in the variable PF.

in Caché TERM. Probably you meant to test it under TUN PLUS 2016? Thanks.

File: [WebTerminal-v4.1.4.zip](https://github.com/intersystems-ru/webterminal/files/723601/WebTerminal-v4.1.4.zip)

**Comment by jcorbalan** [comment]
Does not work, the program shows the characters of the function key that you have pressed, for example if you press F2, it is written "OQ", and nothing should be written

**Comment by ZitRos** [comment]
@jcorbalan, I need to figure out how to test this.

> READ E9#10
> SET PF=$S($T=0:1,$ZB?.N:$ZB\256-31,$E($ZB,4)'=""&($A($ZB,4)'=126):$S($A($ZB,4)-49<5:$A($ZB,4)-39,1:$A($ZB,4)-49),$E($ZB,3)'="":$A($ZB,3)-79,1:-31)

This code does not work in Caché TERM, is it supposed to work there? Or should I use Putty/TUN PLUS 2016/etc to test how it works?

**Comment by jcorbalan** [comment]
This code works inside a routine, not on the command line

 READ E9#10
 SET PF=$SELECT($TEST=0:1,$ZB?.N:$ZB\256-31,$EXTRACT($ZB,4)'=""&($ASCII($ZB,4)'=126):$SELECT($ASCII($ZB,4)-49<5:$ASCII($ZB,4)-39,1:$ASCII($ZB,4)-49),$EXTRACT($ZB,3)'="":$ASCII($ZB,3)-79,1:-31)

**Comment by ZitRos** [comment]
@jcorbalan, it works as next:

![2017-01-23_171603](https://cloud.githubusercontent.com/assets/4989256/22209580/aed212b8-e18f-11e6-8718-5bd22fff2fa4.png)

It doesn't return nothing but `1` for all function keys despite you have stated:

> You can use this command in TERM:, to get the value of the PF key pressed, which will be stored in the variable PF.

**Comment by jcorbalan** [comment]
Please try this code

DUCK
 I $ZV["M/SQL"!($ZV["ISM")!($ZV["Open M")!($ZV["Cache")  W "Press F' Keys: "
 READ E9#10
 SET (PF,%PF5)=$SELECT($TEST=0:1,$ZB?.N:$ZB\256-31,$EXTRACT($ZB,4)'=""&($ASCII($ZB,4)'=126):$SELECT($ASCII($ZB,4)-49<5:$ASCII($ZB,4)-39,1:$ASCII($ZB,4)-49),$EXTRACT($ZB,3)'="":$ASCII($ZB,3)-79,1:-31) 
 write !, "PF: ", PF,!
 Q

**Comment by ZitRos** [comment]
Thank you!

I have found the problem: variable `$ZB` not changing in WebTerminal, whether in Caché TERM it changes:

```
USER>read A#9 zw $ZB
$c(27)_"OS"
```
(Caché TERM)

Let me find out how to make it work...
